src/pixelator/pixeldataset/precomputed_layouts.py

```diff
@@ -33,8 +33,169 @@ class PreComputedLayoutsEmpty(PixelatorBaseException):
     """Raised when trying to access an empty PreComputedLayouts instance."""
 
 
+class RequestedLayoutsHaveMultipleDimensionalities(PixelatorBaseException):
+    """Raised when trying to materialize PreComputedLayouts with multiple dimensionalities."""
+
+
+class _DataProvider(Protocol):
```

Comment:
__init__ method here not needed compare to `_EmptyDataProvider`?

Reply:
Nope.